tensorboard/backend/event_processing/data_provider_test.py

```diff
@@ -0,0 +1,162 @@
+
+from tensorboard.backend.event_processing import data_provider
+from tensorboard.backend.event_processing import (
+  plugin_event_multiplexer as event_multiplexer,
+)
+from tensorboard.compat.proto import summary_pb2
+from tensorboard.data import provider as base_provider
+from tensorboard.plugins.image import metadata as image_metadata
+from tensorboard.plugins.image import summary_v2 as image_summary
+from tensorboard.plugins.scalar import metadata as scalar_metadata
+from tensorboard.plugins.scalar import summary_v2 as scalar_summary
+import tensorflow.compat.v2 as tf
+
+
+class MultiplexerDataProviderTest(tf.test.TestCase):
+  def setUp(self):
+    super(MultiplexerDataProviderTest, self).setUp()
+    self.logdir = self.get_temp_dir()
+
+    logdir = os.path.join(self.logdir, "polynomials")
+    with tf.summary.create_file_writer(logdir).as_default():
+      for i in xrange(10):
+        scalar_summary.scalar("square", i ** 2, step=2 * i)
+        scalar_summary.scalar("cube", i ** 3, step=3 * i)
+
+    logdir = os.path.join(self.logdir, "waves")
+    with tf.summary.create_file_writer(logdir).as_default():
+      for i in xrange(10):
+        scalar_summary.scalar("sine", tf.sin(float(i)), step=i)
+        scalar_summary.scalar("square", tf.sign(tf.sin(float(i))), step=i)
+        # Summary with rank-0 data but not owned by the scalars plugin.
+        metadata = summary_pb2.SummaryMetadata()
+        metadata.plugin_data.plugin_name = "marigraphs"
+        tf.summary.write("high_tide", tensor=i, step=i, metadata=metadata)
+
+    logdir = os.path.join(self.logdir, "pictures")
+    with tf.summary.create_file_writer(logdir).as_default():
+      purple = tf.constant([[[255, 0, 255]]], dtype=tf.uint8)
+      for i in xrange(1, 11):
+        image_summary.image("purple", [tf.tile(purple, [i, i, 1])], step=i)
+
+  def create_provider(self):
+    multiplexer = event_multiplexer.EventMultiplexer()
+    multiplexer.AddRunsFromDirectory(self.logdir)
+    multiplexer.Reload()
+    return data_provider.MultiplexerDataProvider(multiplexer)
+
+  def test_list_scalars_all(self):
+    provider = self.create_provider()
+    result = provider.list_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=None,
+    )
+    self.assertItemsEqual(result.keys(), ["polynomials", "waves"])
+    self.assertItemsEqual(result["polynomials"].keys(), ["square", "cube"])
+    self.assertItemsEqual(result["waves"].keys(), ["square", "sine"])
+    sample = result["polynomials"]["square"]
+    self.assertIsInstance(sample, base_provider.ScalarMetadata)
+    self.assertEqual(sample.max_step, 18)
+    # nothing to test for wall time, as it can't be mocked out
+    self.assertIsInstance(sample.summary_metadata, summary_pb2.SummaryMetadata)
+    self.assertEqual(
+        sample.summary_metadata.plugin_data.plugin_name,
+        scalar_metadata.PLUGIN_NAME,
+    )
+
+  def test_list_scalars_filters(self):
```

Comment:
Maybe we should also include a check here for a filter that specifies both runs and tags?

Reply:
Done.